src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -467,40 +464,38 @@ public static void Reverse(ref int buf, nuint length)
                     tempLast = Vector128.Shuffle(tempLast, Vector128.Create(3, 2, 1, 0));
 
                     // Store the reversed vectors
-                    tempLast.StoreUnsafe(ref first);
-                    tempFirst.StoreUnsafe(ref last);
+                    tempLast.StoreUnsafe(ref buf, (nuint)offset);
+                    tempFirst.StoreUnsafe(ref buf, (nuint)lastOffset);
 
-                    first = ref Unsafe.Add(ref first, Vector128<int>.Count);
-                    last = ref Unsafe.Subtract(ref last, Vector128<int>.Count);
-                } while (!Unsafe.IsAddressGreaterThan(ref first, ref last));
+                    offset += Vector128<int>.Count;
+                    lastOffset -= Vector128<int>.Count;
+                } while (lastOffset >= offset);
 
-                // Shift instead of divide is fine here since we don't care about what the negative numbers are
-                remainder = Unsafe.ByteOffset(ref first, ref Unsafe.Add(ref last, Vector128<int>.Count)) >> 2;
+                remainder = lastOffset + Vector128<int>.Count - offset;
             }
 
             // Store any remaining values one-by-one
             if (remainder > 1)
             {
-                ReverseInner(ref first, (nuint)remainder);
+                ReverseInner(ref Unsafe.Add(ref buf, offset), (nuint)remainder);
```

Comment:
The refactor improves the codegen:

<details>

## Main, X64 RyuJIT AVX2
```assembly
; System.SpanHelpers.Reverse(Int32 ByRef, UIntPtr)
       vzeroupper
       cmp       rdx,10
       jb        short M00_L02
       nop       dword ptr [rax]
       mov       rax,rdx
       shr       rax,3
       shr       rax,1
       vmovupd   ymm0,[7FFEC5FDFFE0]
       xor       r8d,r8d
       test      rax,rax
       jbe       short M00_L01
       nop       dword ptr [rax+rax]
M00_L00:
       lea       r9,[r8*8]
       inc       r8
       lea       r10,[r8*8]
       mov       r11,rdx
       sub       r11,r10
       vmovdqu   ymm1,ymmword ptr [rcx+r9*4]
       vmovdqu   ymm2,ymmword ptr [rcx+r11*4]
       vpermd    ymm1,ymm0,ymm1
       vpermd    ymm2,ymm0,ymm2
       vmovdqu   ymmword ptr [rcx+r9*4],ymm2
       vmovdqu   ymmword ptr [rcx+r11*4],ymm1
       cmp       r8,rax
       jb        short M00_L00
M00_L01:
       shl       rax,3
       lea       rcx,[rcx+rax*4]
       add       rax,rax
       sub       rdx,rax
       jmp       short M00_L05
M00_L02:
       cmp       rdx,8
       jb        short M00_L05
       mov       rax,rdx
       shr       rax,2
       shr       rax,1
       xor       r8d,r8d
       test      rax,rax
       jbe       short M00_L04
M00_L03:
       lea       r9,[r8*4]
       inc       r8
       lea       r10,[r8*4]
       mov       r11,rdx
       sub       r11,r10
       vmovdqu   xmm0,xmmword ptr [rcx+r9*4]
       vmovdqu   xmm1,xmmword ptr [rcx+r11*4]
       vpshufd   xmm0,xmm0,1B
       vpshufd   xmm1,xmm1,1B
       vmovdqu   xmmword ptr [rcx+r9*4],xmm1
       vmovdqu   xmmword ptr [rcx+r11*4],xmm0
       cmp       r8,rax
       jb        short M00_L03
M00_L04:
       shl       rax,2
       lea       rcx,[rcx+rax*4]
       add       rax,rax
       sub       rdx,rax
M00_L05:
       cmp       rdx,1
       jbe       short M00_L07
       movsxd    rax,edx
       lea       rax,[rcx+rax*4-4]
       nop       word ptr [rax+rax]
M00_L06:
       mov       edx,[rcx]
       mov       r8d,[rax]
       mov       [rcx],r8d
       mov       [rax],edx
       add       rcx,4
       add       rax,0FFFFFFFFFFFFFFFC
       cmp       rcx,rax
       jb        short M00_L06
M00_L07:
       vzeroupper
       ret
; Total bytes of code 283
```

## Before, X64 RyuJIT AVX2
```assembly
; System.SpanHelpers.Reverse(Int32 ByRef, UIntPtr)
       vzeroupper
       lea       rax,[rcx+rdx*4]
       cmp       rdx,8
       jb        short M00_L01
       add       rax,0FFFFFFFFFFFFFFE0
       vmovupd   ymm0,[7FFEC68C0500]
       nop       word ptr [rax+rax]
M00_L00:
       vmovdqu   ymm1,ymmword ptr [rcx]
       vmovdqu   ymm2,ymmword ptr [rax]
       vpermd    ymm1,ymm0,ymm1
       vpermd    ymm2,ymm0,ymm2
       vmovdqu   ymmword ptr [rcx],ymm2
       vmovdqu   ymmword ptr [rax],ymm1
       add       rcx,20
       add       rax,0FFFFFFFFFFFFFFE0
       cmp       rcx,rax
       jb        short M00_L00
       lea       rdx,[rax+20]
       sub       rdx,rcx
       mov       rax,rdx
       sar       rax,3F
       and       rax,3
       add       rdx,rax
       sar       rdx,2
       jmp       short M00_L03
M00_L01:
       cmp       rdx,4
       jb        short M00_L03
       add       rax,0FFFFFFFFFFFFFFF0
M00_L02:
       vmovdqu   xmm0,xmmword ptr [rcx]
       vmovdqu   xmm1,xmmword ptr [rax]
       vpshufd   xmm0,xmm0,1B
       vpshufd   xmm1,xmm1,1B
       vmovdqu   xmmword ptr [rcx],xmm1
       vmovdqu   xmmword ptr [rax],xmm0
       add       rcx,10
       add       rax,0FFFFFFFFFFFFFFF0
       cmp       rcx,rax
       jb        short M00_L02
       lea       rdx,[rax+10]
       sub       rdx,rcx
       mov       rax,rdx
       sar       rax,3F
       and       rax,3
       add       rdx,rax
       sar       rdx,2
M00_L03:
       cmp       rdx,1
       jle       short M00_L05
       mov       rax,rcx
       lea       rdx,[rax+rdx*4-4]
M00_L04:
       mov       ecx,[rax]
       mov       r8d,[rdx]
       mov       [rax],r8d
       mov       [rdx],ecx
       add       rax,4
       add       rdx,0FFFFFFFFFFFFFFFC
       cmp       rax,rdx
       jb        short M00_L04
M00_L05:
       vzeroupper
       ret
; Total bytes of code 213
```

## After, X64 RyuJIT AVX2
```assembly
; System.SpanHelpers.Reverse(Int32 ByRef, UIntPtr)
       vzeroupper
       xor       eax,eax
       cmp       rdx,8
       jl        short M00_L01
       add       rdx,0FFFFFFFFFFFFFFF8
       vmovupd   ymm0,[7FFEC6512680]
M00_L00:
       vmovdqu   ymm1,ymmword ptr [rcx+rax*4]
       vmovdqu   ymm2,ymmword ptr [rcx+rdx*4]
       vpermd    ymm1,ymm0,ymm1
       vpermd    ymm2,ymm0,ymm2
       vmovdqu   ymmword ptr [rcx+rax*4],ymm2
       vmovdqu   ymmword ptr [rcx+rdx*4],ymm1
       add       rax,8
       add       rdx,0FFFFFFFFFFFFFFF8
       cmp       rdx,rax
       jge       short M00_L00
       add       rdx,8
       sub       rdx,rax
       jmp       short M00_L03
       nop       word ptr [rax+rax]
M00_L01:
       cmp       rdx,4
       jl        short M00_L03
       add       rdx,0FFFFFFFFFFFFFFFC
M00_L02:
       vmovdqu   xmm0,xmmword ptr [rcx+rax*4]
       vmovdqu   xmm1,xmmword ptr [rcx+rdx*4]
       vpshufd   xmm0,xmm0,1B
       vpshufd   xmm1,xmm1,1B
       vmovdqu   xmmword ptr [rcx+rax*4],xmm1
       vmovdqu   xmmword ptr [rcx+rdx*4],xmm0
       add       rax,4
       add       rdx,0FFFFFFFFFFFFFFFC
       cmp       rdx,rax
       jge       short M00_L02
       add       rdx,4
       sub       rdx,rax
M00_L03:
       cmp       rdx,1
       jle       short M00_L05
       lea       rax,[rcx+rax*4]
       lea       rdx,[rax+rdx*4-4]
M00_L04:
       mov       ecx,[rax]
       mov       r8d,[rdx]
       mov       [rax],r8d
       mov       [rdx],ecx
       add       rax,4
       add       rdx,0FFFFFFFFFFFFFFFC
       cmp       rax,rdx
       jb        short M00_L04
M00_L05:
       vzeroupper
       ret
; Total bytes of code 188
```

</details>

And slightly the perf 👍